modules/apps/27-interchain-accounts/ibc_module_test.go

```diff
@@ -0,0 +1,421 @@
+	chainB *ibctesting.TestChain
+}
+
+func TestICATestSuite(t *testing.T) {
+	suite.Run(t, new(InterchainAccountsTestSuite))
+}
+
+func (suite *InterchainAccountsTestSuite) SetupTest() {
+	suite.coordinator = ibctesting.NewCoordinator(suite.T(), 3)
+	suite.chainA = suite.coordinator.GetChain(ibctesting.GetChainID(0))
+	suite.chainB = suite.coordinator.GetChain(ibctesting.GetChainID(1))
+}
+
+func NewICAPath(chainA, chainB *ibctesting.TestChain) *ibctesting.Path {
+	path := ibctesting.NewPath(chainA, chainB)
+	path.EndpointA.ChannelConfig.PortID = types.PortID
+	path.EndpointB.ChannelConfig.PortID = types.PortID
+	path.EndpointA.ChannelConfig.Order = channeltypes.ORDERED
+	path.EndpointB.ChannelConfig.Order = channeltypes.ORDERED
+	path.EndpointA.ChannelConfig.Version = types.VersionPrefix
+	path.EndpointB.ChannelConfig.Version = TestVersion
+
+	return path
+}
+
+func InitInterchainAccount(endpoint *ibctesting.Endpoint, owner string) error {
+	portID, err := types.GeneratePortID(owner, endpoint.ConnectionID, endpoint.Counterparty.ConnectionID)
+	if err != nil {
+		return err
+	}
+	channelSequence := endpoint.Chain.App.GetIBCKeeper().ChannelKeeper.GetNextChannelSequence(endpoint.Chain.GetContext())
+
+	if err := endpoint.Chain.GetSimApp().ICAKeeper.InitInterchainAccount(endpoint.Chain.GetContext(), endpoint.ConnectionID, endpoint.Counterparty.ConnectionID, owner); err != nil {
+		return err
+	}
+
+	// commit state changes for proof verification
+	endpoint.Chain.App.Commit()
+	endpoint.Chain.NextBlock()
+
+	// update port/channel ids
+	endpoint.ChannelID = channeltypes.FormatChannelIdentifier(channelSequence)
+	endpoint.ChannelConfig.PortID = portID
+	return nil
+}
+
+// SetupICAPath invokes the InterchainAccounts entrypoint and subsequent channel handshake handlers
+func SetupICAPath(path *ibctesting.Path, owner string) error {
+	if err := InitInterchainAccount(path.EndpointA, owner); err != nil {
+		return err
+	}
+
+	if err := path.EndpointB.ChanOpenTry(); err != nil {
+		return err
+	}
+
+	if err := path.EndpointA.ChanOpenAck(); err != nil {
+		return err
+	}
+
+	if err := path.EndpointB.ChanOpenConfirm(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (suite *InterchainAccountsTestSuite) TestOnChanOpenInit() {
+	var (
+		channel *channeltypes.Channel
+	)
+	testCases := []struct {
+		name     string
+		malleate func()
+		expPass  bool
+	}{
+		{
+			"success", func() {}, true,
+		},
+		{
+			"ICA OnChanOpenInit fails - UNORDERED channel", func() {
+				channel.Ordering = channeltypes.UNORDERED
+			}, false,
+		},
+		{
+			"ICA auth module callback fails", func() {
+				suite.chainA.GetSimApp().ICAAuthModule.IBCApp.OnChanOpenInit = func(ctx sdk.Context, order channeltypes.Order, connectionHops []string,
+					portID, channelID string, chanCap *capabilitytypes.Capability,
+					counterparty channeltypes.Counterparty, version string,
+				) error {
+					return fmt.Errorf("mock ica auth fails")
+				}
+			}, false,
```

Comment:
test handshake in wrong direction fails. ie INIT from host chain

Reply:
I agree with the intention, but I think this is out of scope of this pr and I don't think these tests are the place to test this functionality. This pr is replacing the existing hook mechanism with controller side middleware. Testing that the handshake has the correct checks in place to reject the wrong flow isn't what this pr aims to do. Those tests are missing, but the scope of this pr shouldn't be increased to fix that issue

`keeper/handshake_test.go` should test that we have the right checks in place, but we can still add higher level tests. Trying to test wrong flow in these test cases isn't possible, I'd have to add a completely new test (as I'd need to reconstruct the arguments being passed into the callback to make sense - otherwise it'd trivially fail). 

I've opened #522

Reply:
I think it's fine to do a follow-up PR :handshake: